design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,214 @@
+multiple places
+
+### Non-Goals
+
+- Update any metrics or alerts
+- Apply these ideas to events in this phase
+- Create a new component to run in the cluster alongside other components to
+collect metrics and create alerts
+
+## How
+
+### General relationship diagram:
+
+<img src="relationship.png" alt="relationship diagram" width="800"/>
+
+With this refactor, we intend to move out monitoring logic from all the KubeVirt
+components' business logic code. As we see in the above diagram, each KubeVirt
+component will now import a package in the `kubevirt/monitoring` repository.
+Each package will contain, for the corresponding component, the declaration and
+registration of metrics, the generation of the `PrometheusRule` resource with
+the alerts and recording rules, and the automatically generated documentation
+information.
+
+In turn, each component package in `kubevirt/monitoring`, will use a generic
+package for operator observability, abstracting all common operations between
+the components (such as registering them in `controller-runtime` metrics
+`Registry`), deduplicating code, and providing a way to define the metrics,
+recording rules , and alerts in a declarative manner.
+
+### Relationship diagram for each component monitoring code:
+
+<img src="component_monitoring_relationship.png" alt="relationship diagram" width="800"/>
+
+Each component package in `kubevirt/monitoring`, will have a `metrics/` package
+with the metrics declarations, the methods to update/consult the metrics values
+and the collectors logic. The `rules/` package will also have the declarative
+configuration of the resources, in this case, alerts and recording rules.
+
+Finally, the `docs/` package will list the metrics, alerts, and recording rules
+and generate the documentation files using the operator observability docs
+utilities.
+
+### Declarative configuration
+
+The declarative configuration will be done using a YAML files. The YAML files
+will be parsed and used to create the metrics, alerts, and recording rules.
+Components should be able to define their resources in one or more YAML files.
+
+As a general rule, the files will be structured in the following way:
+
+```yaml
+observability:
+  common_labels:
+    - name: controller
+      value: example-operator
+
+  groups:
+    - name: operator_resources
+      common_labels:
+        - name: grouped
+          value: this_is_grouped
+      metrics:
+        - name: example_operator_reconcile_count
+          help: Number of times the operator has executed the reconcile loop
+          type: counter
+          stability: stable
+          labels:
+            - name: another_label
+              value: only_for_this_metric
+
+      recording_rules:
+        - name: example_operator_number_of_pods
+          help: Number of example operator pods in the cluster
+          type: gauge
+          expr: sum(up{namespace='$NAMESPACE', pod=~'example-operator-.*'}) or vector(0)
```

Comment:
I think that we can take this a step forward by presenting common promQL generators that would be declarative and help the user define these without having to deal with promQL itself.

Some thoughts that come into mind are, for metrics:
`expr: ObjectField("spec.age")`
`expr: NumberOfInstances("my-namespace", "virt-api-*")`
`expr: ReadyPods("my-namespace", "virt-api-*")`

for alerts:
`expr: ObjectField("spec.age").IsNumerically(">", 123)`
`expr: DeploymentInstances("deploymentName").IsZero()`

As I see it, `operator-availability` could have pre-made promQL generators (or however we want to call them) that are relevant to most operators out there. Then, every operator (e.g. Kubevirt) would be able to expand them with generators specific to their scope. As an example, for Kubevirt, we can introduce `NodeWithKVM()`, `LiveMigration()`, etc.

I think that with these in place this repo might be very appealing to many other operators as it eases the monitoring effort into a level that it's accessible by developers that aren't familiar with PromQL at all.

WDYT?

Reply:
This doesn't make sense to me. I think we should keep the PromQL language.
If someone is adding a metric they should know PromQL. This is a must.

Reply:
Can you provide a reasoning for your opinion?

Why is it wrong to wrap very basic operations with functions that would make it easier? Why does a user that wants to add a straight-forward metric needs to struggle with promQL code? Why do we need to duplicate the PromQL logic between many different metrics?

I just don't understand what your concerns are / what's the value of requiring to write explicit promQL code.

Reply:
Its like adding a layer over SQL language. It doesn't make sense.

Reply:
Hi. I think that creating a language that is rendered to a query language is a bit overkill here, although if someone manage to create such thing and it plays nicer than promQL itself it should be suggested to prometheus folks as well :)
But instead, having some subset of Go funcs that will return common PromQL expressions could be wonderful.

Reply:
> I think that we can take this a step forward by presenting common promQL generators that would be declarative and help the user define these without having to deal with promQL itself.
> 
> Some thoughts that come into mind are, for metrics: `expr: ObjectField("spec.age")` `expr: NumberOfInstances("my-namespace", "virt-api-*")` `expr: ReadyPods("my-namespace", "virt-api-*")`
> 
> for alerts: `expr: ObjectField("spec.age").IsNumerically(">", 123)` `expr: DeploymentInstances("deploymentName").IsZero()`
> 
> As I see it, `operator-availability` could have pre-made promQL generators (or however we want to call them) that are relevant to most operators out there. Then, every operator (e.g. Kubevirt) would be able to expand them with generators specific to their scope. As an example, for Kubevirt, we can introduce `NodeWithKVM()`, `LiveMigration()`, etc.
> 
> I think that with these in place this repo might be very appealing to many other operators as it eases the monitoring effort into a level that it's accessible by developers that aren't familiar with PromQL at all.
> 
> WDYT?

I don't think its needed for metrics instrumentation. The way to write metrics using Prometheus Go client is pretty straight-forward.

Reply:
When you create a query you need to mine the filtering of the metric (namespace/name/pod/etc.) , what you sum/avg/max/min by and the metric type (Gauge/Counter/Histogram/Summery) will require different functions for doing the same thing.
For example: 
sum(rate(metric_name[time period])) (for counter) vs 
sum_over_time(metric_name[time period]) (for gauge) or perhaps increase(for counters).
We should use the plain promQL and not add extra complexity to it.

Reply:
Obviously I didn't mean to create another layer of promQL or extend the language, but rather introduce helper functions for very frequent operations. I have a feeling that many metrics would be very similar to one another (not only just in Kubevirt, but in many other repos as well) and if that's the case I don't see a reason why not to provide a way of simplification and avoiding code duplication.

I think it's somewhat similar to how we're using gomega. We could have written:
```go
if value == nil {
    return err
}
```

But since it's a very common operation, we rather use something like:
```go
Expect(value).NotTo(BeNil())
```

I wouldn't say that this is an extra layer to Go, just helper functions for frequent operations.

BTW, @machadovilaca's example [here](https://github.com/machadovilaca/operator-observability/blob/main/examples/metrics/custom_resource_collector.go#L16) is somewhat similar. There, a function is being called to count the number of k8s objects in a generic way that doesn't require any promQL knowledge.

I think the advantages are:
1. Simplicity
2. Less error-prone
3. Easier to add straight-forward metrics (which are pretty common) by people who aren't comfortable with PromQL.

I'm also not convinced by the argument that it would be a burden to maintain, since again I'm talking about the most common and straight forward promQL code anyways. The argument for being more error-prone is not compelling to me either, since it's much easier to make a typo in a promQL language IMO.

Reply:
In any case, this discussion should block @machadovilaca from merging this PR.
We can asses this together and can tackle this in the future. Perhaps the right place to discuss this in within an issue in the operator-observability repo

Reply:
This should not be a goal for this feature.
If you (In general) think of things that can help you can create a PR for them.
But it should not be part of the scope for the code refactoring which aims to separate the monitoring code from the core operator code.
We need to scope the effort for this feature. This is a nice to have improvement at should be separated out of this feature please.